[Expressions] add color_rgbf/color_cmykf/color_hsvf/color_hslf

troopa81 opened the pull request:
According to QEP #283 , I propose to add 2 new expression functions to load RGB and CMYK color

Contrary to color_cmyk/rgb, qcolor_cmyk/rgb allows to load a color from float component (and not int) into a QColor object (and not a string representation).

This allows to avoid color conversions so we can write PDF with native CMYK colors coming from expression.

This is a first PR, I'm planning to add other functions with the same goal.

So far, I didn't deprecate the old functions.

**Funded by Bordeaux Métropole**

troopa81 committed:
feat(Expression): Add color_rgbf

Contrary to color_rgb, color_rgbf allows to load a color from float
component (and not int) into a QColor object (and not a string representation).

troopa81 committed:
feat(Expression): Add color_cmykf

Contrary to color_cmyk, color_cmykf allows to load a color from float
component (and not int) into a QColor object (and not a string representation).

nyalldawson commented:
In general I think this is a good idea. One thing I don't particularly like is exposing the name "qcolor" in expressions, as that's an internal detail which has no meaning to the majority of end users. It's a shame we can't just use "color" here, but I agree with your reasoning.

@nirvn @NathanW2 any ideas on alternative names?

nirvn commented:
@troopa81 , have you tried using these new functions with other color functions (e.g. set_color_part , darker, lighter, etc.)? Does it work?

Having looked at this a little bit recently, I think what needs to happen here is for our color functions to both accept QColor objects as parameters as well as return QColor objects, not a string representation. Question here is: can this be done without breaking a bunch of projects out there that assume color functions return a string? Probably not. An alternative would be to introduce a color_v2_* set of functions that would handle colors as QColor, or have an optional parameter that would have preexisting functions return a QColor object, e.g. color_cmyk(1.0, 0.5, 0.0, 0.5, {as_object, as_color, as_true_color, as_whatever :)}:=true).

Which then brings back to this. Do we want to introduce new functions, or should we introduce an optional parameter that dictates whether values are to be handled as 0.0 to 1.0 floats. E.g., color_cmyk(1.0, 0.5, 0.0, 0.5, float_values:=true).

nyalldawson commented:
@nirvn 

>  I think what needs to happen here is for our color functions to both accept QColor

Agreed

> as well as return QColor objects, not a string representation.

I think this is too risky

> color_v2_* set of functions that would handle colors as QColor, or have an optional parameter that would have preexisting functions return a QColor object, e.g. color_cmyk(1.0, 0.5, 0.0, 0.5, {as_object, as_color, as_true_color, as_whatever :)}:=true).

Neither of these are great options in my opinion :disappointed: 

I think for the foreseeable future we're likely to be left with a mix of functions which return strings and qcolors, and just have to ensure that we're always tolerant when READING color values.

nirvn commented:
@nyalldawson , foreseeable future = QGIS < 4.0 right?

If we stick to introducing a new pair of functions here, we're still in need to come up with a better name as I do agree using qcolor as a reference to returning a QColor object is a bit too obscure for users.

color_{rgb,cmyk,rgba,etc.}_float ?

troopa81 commented:
> I think what needs to happen here is for our color functions to both accept QColor

Yes, that's my plan but that works only for:
- create_ramp
- set_color_part

which let us all of these method to duplicate:
- color_grayscale_average
 - color_hsl
- color_hsv
- color_mix_rgb
- color_part
- darker
- lighter
- project_color
- ramp_color

> >   as well as return QColor objects, not a string representation.
> I think this is too risky

I agree, there is a high chance of breaking things without knowing it here

> Which then brings back to this. Do we want to introduce new functions, or should we introduce an optional parameter that dictates whether values are to be handled as 0.0 to 1.0 floats.

I decided to keep range between 0-255 for rgb and 0-100 for cmyk even with float methods. Do we prefer to keep a uniform 0-1 range for the float methods, like Qt ? It's maybe better because it avoid dividing by 255/100 and so losing precision.

> color_{rgb,cmyk,rgba,etc.}_float ?

color_{rgb,cmyk,rgba,etc.}_object ?

nyalldawson commented:
@troopa81 
> which let us all of these method to duplicate:

Actually we don't need to duplicate them at all, just make them more permissive  / flexible.

I would:

1. Add `QgsExpressionUtils::getColorValue`, following the pattern of the existing getDateTimeValue / getXXXValue static functions
2. Ensure we use this everywhere in expressions instead of directly calling `QgsSymbolLayerUtils::decodeColor`
3. Make `getColorValue` check the variant type first, and if it's a QColor, return it directly. If not, fallback to the existing string handling of calling `QgsSymbolLayerUtils::decodeColor`

Existing use of the functions won't be changed, but they'll all then transparently handle QColor objects too.

> color_{rgb,cmyk,rgba,etc.}_float ?

This sounds good to me. I like it better than color_{rgb,cmyk,rgba,etc.}_object because it's clear that it's used when passing float values. And if at some stage in future we can safely change eg color_rgb to return QColor instead of string, then we aren't left with some awkward historic naming mess with `color_rgb( 0 - 255, ... ) -> QColor` vs `color_rgb_object( 0.0 - 1.0, ... ) -> QColor`

Alternatively I'd also be fine with eg `color_rgbf`, `color_cmykf` etc.

troopa81 commented:
> Actually we don't need to duplicate them at all, just make them more permissive / flexible.

What you describe then works only for `create_ramp` and `set_color_part` which takes QColor as arguments but don't return color. It doesn't work for all the other ones. For instance darker could indeed support a QColor or a string as argument but HAS to return a string because other caller of the this method could then call a `substr` on the result for instance. And so we need a duplicate method, one which return a string, another one which return a QColor.

Hence the will to call them _object instead of _float because it's not just about the float, it's about the fact that they return an object, not a string.

nyalldawson commented:
@troopa81 I'm not proposing we change the return value of any existing functions, just the acceptable argument types. And then we take care that all NEW colour functions return objects, never strings.

It's not perfect, but it's the best we can do with the current situation

nyalldawson commented:
The other thing we could do would be to make getStringValue check for qcolor variant values and return the string representation too. That'd ensure that things like substr could always be called on the new functions which return objects...

troopa81 commented:
> @troopa81 I'm not proposing we change the return value of any existing functions, just the acceptable argument types. And then we take care that all NEW colour functions return objects, never strings.

OK, why not. But it won't be possible to call darker/lighter method, or event project_color and stay in the CMYK color representation, they would be automatically converted to RGB when passing to string representation. Maybe no users care, I don't know. We can stay that way and see if anyone needs those, I don't have strong opinion.

> The other thing we could do would be to make getStringValue check for qcolor variant values and return the string representation too. That'd ensure that things like substr could always be called on the new functions which return objects...

Yes, I thought about that, it looks like the cleaner solution. But It would change the returned type of an expression that just call for instance `project_color('my_color')`. The caller (attribute table, layout or even a python plugin) would have to deal with that type change. I'm afraid about the things that it could break. Isn't it an API break?

nyalldawson commented:
> But it won't be possible to call darker/lighter method, or event project_color and stay in the CMYK color representation,

One option could be that we adapt the logic so that if a color object is used as an argument, then the return value is a color object. That'd help with eg the darker/lighter use case.

project_color is trickier. We could maybe get away with just always returning qcolor objects for CMYK colours.

>>make getStringValue check for qcolor variant values and return the string representation too.
> But It would change the returned type of an expression that just call for instance project_color('my_color')

No, it wouldn't. It'd only affect the input values for functions, not the return values.


One other idea, potentially trickier but arguably better: we add a "expression engine version" property to QgsProject. Old projects have version 1, a brand new project gets version 2. Version 2 means colours are always objects. Although this won't help when a user has symbols in a style or QML or some other saved version which uses expressions... 😂

troopa81 commented:
> One option could be that we adapt the logic so that if a color object is used as an argument, then the return value is a color object. That'd help with eg the darker/lighter use case.

Good idea

> project_color is trickier. We could maybe get away with just always returning qcolor objects for CMYK colours.

Having different returned type depending on the value of your input parameter is not fantastic IMHO. I'd rather have explicit method name

> One other idea, potentially trickier but arguably better: we add a "expression engine version" property to QgsProject. Old projects have version 1, a brand new project gets version 2. Version 2 means colours are always objects. Although this won't help when a user has symbols in a style or QML or some other saved version which uses expressions... 

That's IMHO beyond the scope on this PR. I think we should have a way to deprecate functions like we do in our python API and be able to propose way to replace old method to users.

troopa81 commented:
I changed method name to color_rgbf/cmykf and use range from 0.0-1.0 instead of 0.0-255.0/100.0.

More PR will come regarding what we discuss in this one.

troopa81 committed:
feat(Expression): Add color_hslf

Contrary to color_hsl, color_hslf allows to load a color from float
component (and not int) into a QColor object (and not a string representation).

troopa81 committed:
feat(Expression): Add color_hsvf

Contrary to color_hsv, color_hsvf allows to load a color from float
component (and not int) into a QColor object (and not a string representation).

troopa81 commented:
This PR includes also color_hsvf and color_hslf methods for the sake of completeness (and also because they share a bunch of code like preview string and color test compare)

nyalldawson commented on src/core/expression/qgsexpressionfunction.cpp:
```diff
@@ -5869,6 +5869,22 @@ static QVariant fcnColorRgb( const QVariantList &values, const QgsExpressionCont
+  if ( ! color.isValid() )
+  {
+    parent->setEvalErrorString( QObject::tr( "Cannot convert '%1:%2:%3:%4' to color" ).arg( red ).arg( green ).arg( blue ).arg( alpha ) );
+    color = QColor( 0, 0, 0 );
```
I'd return QVariant() here instead.

nyalldawson commented on src/core/expression/qgsexpressionfunction.cpp:
```diff
@@ -5869,6 +5869,22 @@ static QVariant fcnColorRgb( const QVariantList &values, const QgsExpressionCont
+  const float green = static_cast<float>( QgsExpressionUtils::getDoubleValue( values.at( 1 ), parent ) );
+  const float blue = static_cast<float>( QgsExpressionUtils::getDoubleValue( values.at( 2 ), parent ) );
+  const float alpha = static_cast<float>( QgsExpressionUtils::getDoubleValue( values.at( 3 ), parent ) );
+  QColor color = QColor::fromRgbF( red, green, blue, alpha );
```
I'd suggest clamping the float values to 0-1, otherwise qt will spam a billion messages to the console and effectively freeze qgis if an out-of-range value is passed to the function

troopa81 replied:
Shall we not detect it and report as an evaluation error, so user can fix its mistake instead of silently creating clamped color?

nirvn replied:
@troopa81 , I'd be -1 to making an out of range value an evaluation error. It can sometimes happen with vector tile layers having complex expressions converted from mapboxgl. I think we're much better being tolerant here, and clamping the value to what's accepted. It'd also just break away from a well established behavior of color function tolerance on that front.

For people who are developing expressions, we can assume they'll be looking for expected color outcomes and so the visual feedback will be the "reporting" here :)

nirvn replied:
@troopa81 , thanks.

troopa81 committed:
fix(Expression): Returns QVariant for invalid color

nyalldawson commented:
@nirvn what's your thoughts here?

troopa81 committed:
fix(Expression): clamp color components

nyalldawson commented:
Thanks @troopa81 !
